cctrl/keyhelpers.py

```diff
@@ -175,3 +176,66 @@ def ask_user_to_use_default_ssh_public_key():
         raise InputErrorException('SecurityQuestionDenied')
 
     return 0
+
+
+def get_public_key_fingerprint(key_path):
+    if not key_path:
+        key_path = get_default_ssh_key_path()
+
+    try:
+        public_key = open(key_path).read()
+    except IOError:
+        return None
+
+    key = base64.b64decode(public_key.strip().split()[1].encode('ascii'))
+    fp_plain = hashlib.md5(key).hexdigest()
+    return ':'.join(a + b for a, b in zip(fp_plain[::2], fp_plain[1::2]))
+
+
+def get_default_ssh_dir():
+    if sys.platform == 'win32':
+        ssh_path = os.path.expanduser('~') + "/.ssh"
+    else:
+        ssh_path = os.getenv("HOME") + "/.ssh"
+
+    return ssh_path
+
+
+def get_default_ssh_key_path():
+    return get_default_ssh_dir() + "/id_rsa.pub"
+
+
+def get_key_from_agent(fingerprint):
+    a = paramiko.agent.Agent()
+    for key in a.get_keys():
+        fp = base64.b16encode(key.get_fingerprint())
+        if fp == fingerprint.replace(':', '').upper():
+            return key
+    return None
+
+
+def sign_token(key_path, fingerprint, data):
+    # from agent
+    pkey = get_key_from_agent(fingerprint)
+    if not pkey:
+        # or from file (without passphrase)
+        # assuming '.pub' file extension
+        if not os.path.exists(key_path[:-4]):
+            raise SignatureException('WrongKeyPath')
+        try:
+            pkey = RSAKey.from_private_key_file(key_path[:-4])
+        except PasswordRequiredException:
+            raise SignatureException('EncryptedKey')
+
+    if not pkey:
+        raise SignatureException('KeyNotFound')
+
+    try:
+        s = str(pkey.sign_ssh_data(data))
```

Comment:
paramiko is inconsistent here in that the agent's key returns `Message` objects for sign_ssh_data and the plain RSAKey returns byte strings. @MatthiasWiesner worked around this by casting to str and then back to `Message`. That's why it should be commented